### PR TITLE
Stake V8, fix PoW and RSA exploits.

### PR DESCRIPTION
**Pull request by tomasbrod**
Status: **merging**

As in email:
I would like to fix the issue with 1) unchecked rsaweight, 2) magnitude
weight multiplication issue and 3) proof of work nonce being included in
pos kernel making this effectively a pow consensus coin.

I think I fixed the above issues by introducing a new stake kernel and
bumping the block version to 8. I named this new kernel "V8 kernel".
This new kernel does not consider rsa and magnitude weight for security
reasons. Does not include nonce but include the Stake Modifier. See some
recent github issues and wiki for explanation why rsa/mag weight is
dangerous.

I based my modification on refactored miner from #301. That pull request
is quite large in diff but it really is just an refactor and I already
tested it successfully. It is possible to manually rebase my fix on
master, but it reauires considerable effort better spent somewhere else.

I would like to test this on TestNet first and get some reviews. I
enabled this new kernel in testnet from block no 271700. Testnet nodes
that don't update until this block comes will ban updated nodes due to
incompatibility.

Can we test this?

V8 Kernel Protocol
Tomas Brod 05.06.2017 (dd.mm.yyy)
Plug proof-of-work exploit.
Stake modifier is included without much understanding. Without the
modifier, attacker can create transactions which output will stake at
desired time. In other words attacker can check wheter transactionoutput
will stake in the future and create transactions accordingly. Thus
including modifier, even not completly researched, increases security.
Edit: I now understand how SatakeModifier works and think it is secure.
Note: Rsa or Magnitude weight not included due to multiplication issue.
Note: Payment age and magnitude restrictions not included as they are
not important in my view and are too restrictive for honest users.

### CODE REVIEW

**Comment by tomasbrod** [comment]
For everyone who sets up a testnet node:
- compile this yourself (clone repo and checkout stakev8 branch)
- use debug=true
- add my node below
- keep the logs

Existing testnet nodes that don't run this code will ban you for producing blocks unknown to them.

I will run 2 nodes. One staking, the other with zero coins. I am behind NAT, but I have set up UPnP forwarding and dns.
```
addnode=mangan.holes.sk:32748
```
```
git add remote brod https://github.com/tomasbrod/Gridcoin-Research
git fetch brod
git merge brod/stakev8
```

**Comment by tomasbrod** [comment]
Thank you @Foggyx420 for reporting. _Vehemently_ rejected, locally generated, v7 blocks are expected as I didn't implement all of the checks in Miner and instead relying on main block checking. This should  happen much less with higher difficulty. 

```
ERROR: CheckBlock[]:  Old testnet client spamming new blocks after mandatory upgrade
```
This one was unexpected and got me by surprise. Somehow the client version is missing the leading "v", which causes wrong conversion to number and triggers mandatory update check. I removed the check in previous commit and I will investigate why the version string changed format from expected `v3.5.9.0-g-research` to `3.5.9.0a-58-g787c519`.

**Comment by iFoggz** [comment]
i had an issue with 3.5.9.0a that compiled once with the commit in version with same effect which made my block version show as 590 instead of 3590. i got another fresh copy and compiled and it worked. so idk.

**Comment by tomasbrod** [comment]
Ok. I found where that other version format comes from, but I haven't figured out why and when it occurs. Hints: build/build.h, share/genbuild.sh, src/version.cpp.
I recompiled my node with the "v" and my staked v7 blocks are accepted. Note that the new v8 blocks will be generated starting with block 271702. So please continue running @Foggyx420.

**Comment by iFoggz** [comment]
ill add the change now thx.

**Comment by denravonska** [comment]
Running my node with this as well. Some blocks are rejected, some aren't.

**Comment by iFoggz** [comment]
yea thats normal atm u can see above about the checks for v7

**Comment by denravonska** [comment]
Is the commit where it also logs to disk when using `-printtoconsole` gone? I'm not seeing anything going into debug.log.

**Comment by iFoggz** [comment]
i thought it was to print to console and debug.log at same time

**Comment by iFoggz** [comment]
vancouver01 is also runing his stake kernel on testnet under a different binary name so my prod binaries wont be touched :) working on toronto01 now. got a /usr/bin/testnetgridcoinresearchd now lol

**Comment by iFoggz** [comment]
toronto up too. should also note right now the staking arrow is not green. but im staking even thou staking says false .

**Comment by denravonska** [comment]
Same for me. Gray arrow with "Searching for mature coins" tooltip but still staking.

**Comment by tomasbrod** [comment]
```
ERROR: GetKernelStakeModifier() : reached best block ...
CalculateStakeHashV8: Unable to get stake modifier
```
The modifier errors are caused by coins younger than 5h 59m staking.
nStakeMinAge is 6 hours so I don't know

**Comment by tomasbrod** [comment]
During testing an issue with the stake modifier was discovered. Sometimes GetKernelStakeModifier fails to find the modifier to use and instead of failing the block it is set to zero. This opens an unknown weakness. Clarification: When the stake modifier is not found, zero is used instead. Even without one it is still more secure than V7.
I am only getting one connection besides my nodes and that is 50.98.203.232 from Canada.
Besides that, blocks are staked as they should.
When the modifier is fixed, we should do one more test run.
Thanks for reporting the other lower-priority issues.

**Comment by iFoggz** [comment]
the .232 is my home node @tomasbrod. like i said i pulled ur repo to compile on here to make it work with proper version but the ones on my 23.* nodes i pulled from development on gridcoin repo. i'll recompile.
good news thou i got 4-5 connections and ive staked numerous blocks so far. had a few conflicted but then after that ive gotten numerous good ones. perhaps it was only conflicted as there was less nodes supporting my claim and more supporting a different block. thats just consensus. i'll fix my nodes after i make a coffee and we will get his going on full test mode. cheers

**Comment by iFoggz** [comment]
i have 
1) 91.127.*.* (inbound)
2) 91.127.*.* (inbound)
3) 79.136.*.* (outbound)
4) 99.27.*.* (outbound)

**Comment by iFoggz** [comment]
Tomas had a connection issue on my end and while its syncing up I noticed its still finding kernels while syncing up. shouldn't that behavior not occur while syncing? 

also compiling development sitll giving me a git commit in version

gridcoin@toronto01:~/.GridcoinResearch/testnet/testnet$ testnetgridcoinresearchd -testnet getinfo
{
    "version" : "3.5.9.0a-54-g5307b1c",

-- also cant get in sync with my local wallet that staked a bunch of blocks.stuck on 272043

**Comment by iFoggz** [comment]
also should note all my confirmed blocks on testnet are not on the main testnet chain.

**Comment by tomasbrod** [comment]
Ok @Foggyx420. Thank you for testing this out. Most of the issues you experienced were due to other nodes pushing their v7 blocks. And generally issues of low network participation. I think I fixed that version issue in this branch (cherry-pick candidate). 

The stake modifier issue is bit more serious and I had to release incompatible change once more. So please, you have to pull and recompile. I rolled back my chain to that of `testnet.grcnode.co.uk` and set new V8 block to start generating at 272272. The node is at new ip, same hostname `mangan.holes.sk:32748` and I decreased my bantime to 5 minutes with `-bantime=300`.

I am not sure if The way I changed the stake modifier is secure. I will explain tomorrow.

**Comment by iFoggz** [comment]
ill get my nodes updated and local wallet. ill have to roll back the blocks. may jsut redownlaod them since it doesnt take long on a 150mbit connection and 100 mbit connectison on my nodes

**Comment by iFoggz** [comment]
ill use ur repo to since the version number is messed up the development branch still. keep getting the commit in version and screwing with my wallet.

**Comment by iFoggz** [comment]
got my two nodes with a mirror of ur repo @tomasbrod still says not staking and getmininginfo says its offline with 6 connections on block 272334. still havent staked any pos blocks on either node.. loading up local wallet now. appears to be moving ahead so idk yet we will see how it pans out in the next while. 

there appears to be a wallet with the messed up version with git commit in it. they have staked every block since 272278 on there own. looks like more issues have arrived

**Comment by iFoggz** [comment]
looks like they staking the minimum 0.01

**Comment by tomasbrod** [comment]
I stake so much because I have 86k coins in a lot of utxos and the difficult was low. The wallet that has commit in version is mine. My cpid is 46f64d69..., so you can see who staked which block. There are blocks from other cpids, one of them is 55cd02be iFoggz.canada. You are staking too, so I would say it is OK.

My wallet version is "3.5.9.0a-63-gd5643f2" which is also in PoR blocks. I disabled the check for client version (only block version matters). I will look into it. Also nodes from development branch will ban nodes from stakev8 branch.

One of Your wallets is stuck, probably it is on wrong fork. It should un-stick after more blocks are created or you can rollback the chain. Forking is expected when the difficulty is low and we don't have much coin support and @denravonska hasn't updated.

For reference my chain is {99df7eb227f015934c1d,  272474,  11.06.2017 13:42:08}.

**Comment by denravonska** [comment]
I have (had) updated but my wallet is super stuck :) Can we assemble a node list here? Mine is ormgas.com:32748

**Comment by tomasbrod** [comment]
One my non-staking node says  it is offline too despite being on the same chain as the staking node. I guess it is an effect of low diff. I restarted the node and it is fine.
If it is on a wrong chain then it will take a while to jump back, you can try to rollback.

For the green up-arrow: The mining status does not seem to be properly updated. However `MiningInfo 5` (in getmininginfo and client message) IS updated and reflects current staking status.

Node-list:
23.111.74.106:32748
23.111.69.149:53292
50.98.203.232:38556
mangan.holes.sk:32748
78.99.65.36:32748

**Comment by iFoggz** [comment]
both my nodes were on same chaina nd it appears they are now on same chain with my local wallet and all have staked my local wallet and vancouver are on same block. my toronto node fell behind about 20 blocks. we need more nodes to help this a long

the top 3 ips u have mentioned are mine

**Comment by tomasbrod** [comment]
Fork-check:
```
272266 b17cebb3a5275c68237ab212895221edf819372c417a0a6060069c7ac50cd3dc V7
272267 e497cc4da005315ac8855218369122c64ed77af8cdc4752c6ee3d5fb30ddf2c6 V7+
272272 29f75247ea9b381a82e55b821fbdbb08a63dab7745fe9e4666e87b0b52c68327 V8
272571 98bedf5da443e394e4f64cbf14d1dc7ba873b991d0e5df2a4a5abdf8463c9c11 V8
272600 3c287b736fbd30abceb59772c5a2dc75ebab7daa7724cc709f77fa9b8231fd00 V8
{SBC} SetBestChain: new best=23abe65b603e3ae12d3b  height=272608  trust=1435791938806784  blocktrust=2970521  date=11.06.2017 19:28:48
```

You can force to switch chain with new `reorganize` command.

**Comment by iFoggz** [comment]
i match the hashes on nodes and local wallet.
--
yea i noticed that one cpid is the one staking most of all the blocks. i have 12K balance on testnet with 2800 mag and last block i staked was 105 blocks ago. about little over 5 hours (probably the maturity time, etc is different in ur test version then testnet -- i do have over 300 txs in my wallet that cud stake thou). Also noticed we tend to stake numerous blocks in a row as well which is not too much different then v7 thou not as consistent as it currently seems to be doing.

We do got to be careful about the harder to stake with lower balance essentially thats like saying the higher balance has the better chance of staking even more often now then before (even without the newb boost etc). this can mean more and more blocks in a row staked for a single user through the estimated ~960 a day we try to aim for.

I don't have a problem with the current staking times thou i do agree that any security holes need to be fixed. The network does need to be as secure as possible. 

if things have changed with this what is the new formula for estimated time to stake in client. I already know its a gross estimate but curious to the difference :). 

Things seem to be smooth so far but still long term testing is needed and i'll leave 2 nodes and my testnet local wallet on ur setup. lets get long term results. :)

**Comment by tomasbrod** [comment]
Interest is encoded on only 2 decimal places in boincblock and 0.0051 is rounded up to 0.01, that is why we can't stake less or equal to 0.005 interest even thought mint limiter is only 0.000010 (depends on diff). Interesting.
Edit: The rounding is also applied when deserializing. So just increasing places won't work. Forking change necessary. Or we can leave it as is.

**Comment by tomasbrod** [comment]
My node started staking blocks with bad cpid (cpid="", cpidv2=md5("")). Last good block {eaf04a9d43c02c24f7d08f90d2d4105e17a43461ae3be0d71507f97cb70ab2aa 272795}. My other node rejected the block with "Bad cpid", but **accepted** it when I restarted it, which is strange. I reorganized to that hash, restarted node and staked good block {9bce8fa83bc8d8b1a70a495c62fcb010b3187e3f750d6dba7ee8f3ce098e0d29 272796}. I will wait for blocks from others before letting it stake again.
I don't know why that happened and why it continued staking more blocks on that chain. This is another issue to look into and I am sure it is not caused by my the v8 kernel.

Edit:  I got some blocks by ravon and my node reorganized correctly on the main chain and is staking right (almost every block!).
{e0615cc0d673db38223c19fa836eb9abf68fc8dec96d7e8a2ee6522a89844779 272850}

**Comment by tomasbrod** [comment]
Another empty cpid block was staked recently, but this time not by me.
{776e21c102bae16570dc48cecf952b20cf0d38e6ffa7c646325f83fb265a5c82 272949}
It was accepted on my nodes, because they were in *InitialBlockDownload*.
The other testnet chain is higher then ours and sometimes their nodes connect to ours. They have more blocks so our nodes think they are behind and enter this mode. CPIDs are not checked when `IsInitialBlockDownload`.

Two issues: why this block was staked, and why CPIDs are not checked. Both of these issues are not caused by v8 kernel. I only changed the kernel, not boinchash/dpor handling.

Edit: It was my node and it created a chain of 108 of these blocks. Before that I got message: `CPUMiner: Gridcoin is downloading blocks Or CPIDs are not yet loaded...`. I reorganized to last valid block {e5db4c779caed7b00c56a61359508221d85c77630c71b7ea2969bd902c03c271 272845}. My node ran different code, so that may be reason.

**Comment by tomasbrod** [comment]
Functions responsible for block checking are: ProcessBlock, CheckBlock, AcceptBlock and ConnectBlock. Under some conditions, some of the checks are skipped and othervise inalid blocks are let into the chain. The checks that are skipped are mostly cpid and research reward related.
In our testing, I screwed up my staking node causing it to stake blocks with empty cpid and cpidv2=md5(""). These blocks were accepted into the chain.
This is worrying issue.

**Comment by iFoggz** [comment]
why did it skip? is it assuming that the other chain is so far ahead that the consensus between those nodes must be right? other wise they would not be synced? this must be a faster way to sync

**Comment by iFoggz** [comment]
so i saw some blank cpids again and i was like i staked those blocks. then i checked again like 5 minutes later and then my cpids were in the hashboinc. lots of reorganizing and what not. i have blocks ive staked in my transactions out of order. 80 82 81 for stakes.., etc.

**Comment by tomasbrod** [comment]
Yeah. My node staked a lot of empty cpid blocks again, your node @Foggyx420 probably rejected at first, but then accepted them causing the reorganize and reject of your blocks. I am abandoning the current chain and shutting my nodes until we can add a check for this.

**Comment by tomasbrod** [comment]
Yeah. My node staked a lot of empty cpid blocks again, your node @Foggyx420 probably rejected at first, but then accepted them causing the reorganize and reject of your blocks. I am abandoning the current chain and shutting my nodes until we can add a check for this.
For @denravonska, I didn't change the block accepting code, this issue is present in production network too. In prod wallet blocks like this are thankfully not created.

**Comment by iFoggz** [comment]
Idk if we need the perpend "v" part as i believe @denravonska solved the version problem the other night. could this make a double v?

good to see progress!

**Comment by tomasbrod** [comment]
Hey guys. I pushed some of the changes to block checking that should prevent empty cpids or other invalid dpor blocks entering the blockchain. The "v" should no longer be an issue as I disabled that check. Invalid blocks already in the chain will be rejected. In prod, this will probably be solved with grandfather rule and checkpoint, but in test, forking several weeks back is acceptable.
The tally will run more often, more cpu load.
Wallet will try to stake even when out of sync... :(

**Comment by denravonska** [comment]
Do you need both Windows and Linux nodes or is Linux enough?

**Comment by tomasbrod** [comment]
Oh, sorry for the force pushes. One windows node to provide NN data would be nice.

**Comment by tomasbrod** [comment]
I am sorry. It does not work. No testing. I should have tried it before calling out.
I fails to load superblock in history validation from a year ago.

**Comment by tomasbrod** [comment]
I am sorry. It does not work. No testing. I should have tried it before calling out.
I fails to load superblock in history validation from a year ago. It was probably a period of no superblock (lookback 14 days).

**Comment by denravonska** [comment]
We are in that state right now. No SB for over 16 days.

**Comment by tomasbrod** [comment]
Side note: you can force tlit to re-processing all blocks with new rules as if received, by renaming `blk0001.dat` to `bootstrap.dat`. This process still takes time, but is faster than re-download

**Comment by tomasbrod** [comment]
The last modification did not work. The idea was to perform all validation on blocks being connected. Especially the magnitude and dpor reward. But during import of historic blocks, the checks failed. One (possibly more) blocks magnitude did not match. Maybe that block is really invalid and slipped into the chain thanks to skipped checks, or the superblock where magnitudes are stored was loaded incorrectly. Remember that it had to read historic superblocks too to validate. Maybe as was said, the superblock was missing for longer and thus failed to load.

**Comment by tomasbrod** [comment]
I know that I convoluted this PR by the commits, so If you want to do something with the previous contents revert to commit f35fe1c14c1135b317a432d1cc642b0c29b6a644.

**Comment by denravonska** [comment]
@tomasbrod Did you try it with @Foggyx420's fix where we increase the SB lookback to 90 days on testnet?

**Comment by tomasbrod** [comment]
To undo the mess, I removed my "big eyes" modifications. I also changed the stake kernel a bit to respect timestamp mask.
The bug that caused failure of previous testing run was plugged by a quick-fix. To properly handle that family of bugs, `BlockNeedsChecked` and vulnerable `OutOfSyncByMoreThan` functions must be removed and larger rewriting will be necessary.
The rest is just minor modifications.
I started testing in private fork.

**Comment by tomasbrod** [comment]
Minor corrections. Testing in another fork.
Added option to include arbitrary text in staked block for identification.
If you want to test, _delete_ txleveldb, _rename_ blk0001.dat to bootstrap.dat, set your wallet nick-name and add the usual node. There is no superblock on that fork, to stake you must switch to _investor_ mode. The node nick name (`-org=` option) is to identify the staker for easier _debugging_. Set it to something like your github nick and unique suffix for each node. Sample config below.

```
org=<YourNick.WalletNick>
addnode=mangan.holes.sk:32748
debug=true
boincdatadir=/tmp/nonexistent/
minersleep=8000
bantime=120
suppressupgrade=true
irc=false
dnsseed=false
```

Added: the first block in this fork is `{4f678be9b9fd7c0f18948f57931700cb041b5ef86c3ed179530e8141938d613b 288160}`.
Added: there is superblock, missed that, you can use normal boinc dir.

**Comment by tomasbrod** [comment]
Test reports, fraction of blocks staked by nodes.
First test, two nodes with roughly same balance:
```
{
  "general" : {
  "blocks" : 1001,
  "first_height" : 288160,
  "last_height" : 289160,
  "time_span_hour" : 28.26666667
 },
 "orgs" : {
  "TomasBrod.Mg.A" : 0.41458541,
  "TomasBrod.Mg.C" : 0.39360639,
  "" : 0.19180819
}}
```
Node MgA got some research rewards that increased it's balance. I guess 41.5% to 39.4% is close enough and within the error margin.

Second test, 3 nodes, MgA with 59kGRC, MgC with 20k and MgD with 10k.
```
{ "general" : {
  "blocks" : 532,
  "first_height" : 290100,
  "last_height" : 290631,
  "time_span_hour" : 14.16000000
 },
 "orgs" : {
  "TomasBrod.Mg.A" : 0.69548872,
  "TomasBrod.MgC$20k" : 0.21052632,
  "TomasBrod.MgD$10k" : 0.09398496
}}
```

Here the error is sightly higher I am not sure what is the cause but I guess it is still acceptable.
The percentages are supposed to be 66%, 22% and 11%.

The average block spacing over the entire testing fork  was 98.8 seconds.
There were on average 874.4 blocks per day, which is little lower than the 1k bpd target.

You can come and test yourself @denravonska and @Foggyx420. But I guess this is good enough.

**Comment by denravonska** [comment]
@tomasbrod Awesome! I guess there's no need for larger scale testing within the PR so we can postpone it until after merge, right?

What more remains to be done on it?

**Comment by tomasbrod** [comment]
Squash/reorder/rebase some commits, merge,
And set block height for kernel switch in prod net as well.

**Comment by tomasbrod** [comment]
I cleaned the patchset, squashed few commits, etc. Orig at d0303a9eed3cddffc490122137eeb9d73650394b.

**Comment by tomasbrod** [comment]
Pushed resolution to conflicts. Hopefully nothing was left out.
~~Dont forget to merge cherry-picked stuff in #421~~. **Ready to merge.**

File: src/kernel.cpp [hunk 1]
```diff
@@ -798,3 +798,157 @@ bool CheckStakeModifierCheckpoints(int nHeight, unsigned int nStakeModifierCheck
...
+// V8 Kernel Protocol
+// Tomas Brod 05.06.2017 (dd.mm.yyy)
+// Plug proof-of-work exploit.
+// TODO: Stake modifier is included without much understanding.
+// Without the modifier, attacker can create transactions which output will
+// stake at desired time. In other words attacker can check wheter transaction
+// output will stake in the future and create transactions accordingly.
+// Thus including modifier, even not completly researched, increases security.
+// Note: Rsa or Magnitude weight not included due to multiplication issue.
+// Note: Payment age and magnitude restrictions not included as they are not
```
Comment (denravonska) [inline src/kernel.cpp]: Without this limit, is there a risk that someone stakes many blocks in a row?

File: src/kernel.cpp [hunk 2]
```diff
@@ -798,3 +798,157 @@ bool CheckStakeModifierCheckpoints(int nHeight, unsigned int nStakeModifierCheck
...
+    CBigNum hashProofOfStake( Hash(ss.begin(), ss.end()) );
+    return hashProofOfStake;
+}
+
+int64_t CalculateStakeWeightV8(
+    const CTransaction &CoinTx, unsigned CoinTxN,
+    const MiningCPID &BoincData)
+{
+    int64_t nValueIn = CoinTx.vout[CoinTxN].nValue;
+    nValueIn /= 1250000;
```
Comment (denravonska) [inline src/kernel.cpp:841]: Is there a reason why this is divided by 1.25 million while the V3 weight is divided by 125k? Won't that produce a tiny weight?

Comment (denravonska) [reply src/kernel.cpp:841]: The weight is applied to the target, and if we go above the target we fail? Does that mean that it's 10 times harder to PoS stake with V8?

Comment (tomasbrod) [reply src/kernel.cpp:841]: No. It is 10 times harder to stake the first 1 - 5 version 8 blocks. Difficulty will re-adjust, as we seen on testnet.

Comment (skcin) [reply src/kernel.cpp:841]: @denravonska I am thinking about this constant for a while now. If I remember correctly other PoS coins don't have it. My understanding is that eventually the target will change accordingly to keep the block time at 90 seconds. But I wonder why it is there if it does nothing.

Comment (tomasbrod) [reply src/kernel.cpp:841]: It prevents the BlockTrust and ChainTrust from growing to big. And imposes lover limit on stakable utxo value.

**Pull request merged**